modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/subscription/AndesSubscriptionManager.java

```diff
@@ -274,6 +275,256 @@ private void removeLocalSubscriptionAndNotify(AndesSubscription subscription) th
+            }
+        }else{
+            if(isExactMatchBindingKey){
+                filteredSubscriptions = getInactiveSubscriptionsByByExactBindingKeyMatch(isDurable,protocolType,
+                        destinationType,bindingKeyPattern,subscriptionIdPattern,isExactMatchSubscriptionId);
+            }else{
+                filteredSubscriptions = getInactiveSubscriptionsByTokenizedBindingKeyMatch(isDurable,protocolType,
+                        destinationType,bindingKeyPattern,subscriptionIdPattern,isExactMatchSubscriptionId);
+            }
+        }
+
+        return filteredSubscriptions;
+    }
+
+    /**
+     * Get active subscriptions filtered by identifier pattern and exact binding key
+     *
+     * @param isDurable true if searching for durable subscriptions
+     * @param protocolType protocol of the subscription
+     * @param destinationType type of subscription (QUEUE/TOPIC/DURABLE_TOPIC)
+     * @param bindingKeyPattern regex to match with binding key of subscriber
+     * @param subscriptionIdPattern regex to match with ID of the subscriber
+     * @param isExactMatchSubscriptionId exact match of subscription id or not
+     * @param connectedNode id of the node to which the subscriber is connected to
+     * @return Set of subscriptions filtered according to search criteria
+     * @throws AndesException
+     */
+    private Set<AndesSubscription> getActiveSubscriptionsByExactBindingKeyMatch(boolean isDurable,
+                                                                               ProtocolType protocolType,
+                                                                               DestinationType destinationType,
+                                                                               String bindingKeyPattern,
+                                                                               String subscriptionIdPattern,
+                                                                               boolean isExactMatchSubscriptionId,
+                                                                               String connectedNode) throws
+            AndesException {
+
+        Set<AndesSubscription> filteredSubscriptions = new HashSet<>();
+        String messageRouter = destinationType.getAndesMessageRouter();
+
+        Query<AndesSubscription> subscriptionQuery = and
+                (equal(AndesSubscription.DURABILITY, isDurable), equal(AndesSubscription.NODE_ID, connectedNode),
+                 equal(AndesSubscription.PROTOCOL, protocolType), equal(AndesSubscription.ROUTER_NAME, messageRouter),
+                 equal(AndesSubscription.ROUTING_KEY, bindingKeyPattern.toLowerCase()));
+
+        Iterable<AndesSubscription> subscriptions = subscriptionRegistry.exucuteQuery(subscriptionQuery);
+
+        for (AndesSubscription subscription : subscriptions) {
+            if(isExactMatchSubscriptionId){
```

Comment:
We are doing this check per subscription. I think we need to move this a level up and have 2 private methods which accept the list of subscriptions for the 2 cases.